core/broadcast.go

```diff
@@ -270,46 +267,73 @@ func (d *dispatcher) stop() {
 }
 
 type sender struct {
-	ctx    context.Context
-	l      log.Logger
-	client net.ProtocolClient
-	to     net.Peer
-	newCh  chan broadcastPacket
+	ctx      context.Context
+	l        log.Logger
+	client   net.ProtocolClient
+	to       net.Peer
+	directCh chan broadcastPacket
+	newCh    chan broadcastPacket
 }
 
 func newSender(ctx context.Context, l log.Logger, client net.ProtocolClient, to net.Peer, queueSize int) *sender {
 	return &sender{
-		ctx:    ctx,
-		l:      l,
-		client: client,
-		to:     to,
-		newCh:  make(chan broadcastPacket, queueSize),
+		ctx:      ctx,
+		l:        l,
+		client:   client,
+		to:       to,
+		directCh: make(chan broadcastPacket, 1),
+		newCh:    make(chan broadcastPacket, queueSize),
 	}
 }
 
-func (s *sender) sendPacket(p broadcastPacket) {
-	select {
-	case s.newCh <- p:
-	default:
-		s.l.Debug("broadcast", "sender queue full", "endpoint", s.to.Address())
+func (s *sender) sendPacket(p broadcastPacket, direct bool) {
+	if direct {
+		s.directCh <- p
+	} else {
+		select {
+		case s.newCh <- p:
+		default:
+			s.l.Debug("broadcast", "sender queue full", "endpoint", s.to.Address())
+		}
 	}
 }
 
 func (s *sender) run() {
-	for newPacket := range s.newCh {
-		s.sendDirect(newPacket)
+	directOpen := true
+	normalOpen := true
+	for directOpen || normalOpen {
+		for directOpen {
```

Comment:
Wait I'm not sure what do we want here... This node will only send its packet first before rebroadcasting the others ?

Reply:
From what I can understand, this (with respect to the previous commit) makes sure no packets gets sent out before I send mine. Whereas before the logic was I send my packet directly as soon as I get it, but I rebroadcast other packets in the meantime as well.
Why is the former logic desirable over the latter ?

Reply:
Given the test is not passing, I'm reverting to previous logic...